hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/MetricsContext.java

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.impl.metrics;
+
+import com.hazelcast.internal.metrics.MetricTagger;
+import com.hazelcast.internal.metrics.MetricsCollectionContext;
+import com.hazelcast.internal.metrics.ProbeLevel;
+import com.hazelcast.internal.metrics.ProbeUnit;
+import com.hazelcast.jet.core.metrics.Metric;
+import com.hazelcast.jet.core.metrics.Unit;
+
+import javax.annotation.Nonnull;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+import java.util.function.BiFunction;
+
+public class MetricsContext {
+
+    private String onlyName;
+    private Metric onlyMetric;
+    private Unit onlyUnit;
+
+    private Map<String, Metric> metrics;
+
+    Metric metric(String name, Unit unit) {
+        return metric(name, unit, SimpleMetric::new);
+    }
+
+    Metric threadSafeMetric(String name, Unit unit) {
+        return metric(name, unit, ThreadSafeMetric::new);
+    }
+
+    private Metric metric(String name, Unit unit, BiFunction<String, Unit, Metric> metricSupplier) {
+        Metric res = null;
+        if (name.equals(onlyName)) {
+            res = onlyMetric;
+        } else if (metrics != null) {
+            res = metrics.get(name);
+        }
+        // register metric on first use
+        if (res == null) {
+            res = metricSupplier.apply(name, unit);
+            if (onlyName == null) {
+                // the first and so far the only metric
+                onlyName = name;
+                onlyMetric = res;
+                onlyUnit = unit;
+            } else {
+                // 2 or more metrics
+                if (metrics == null) {
+                    metrics = new HashMap<>();
+                    metrics.put(onlyName, onlyMetric);
+
+                    onlyName = null;
+                    onlyMetric = null;
+                    onlyUnit = null;
+                }
+                metrics.put(name, res);
+            }
+        }
+        return res;
+    }
+
+    public void collectMetrics(MetricTagger tagger, MetricsCollectionContext context) {
+        if (onlyMetric != null) {
```

Comment:
we should also maybe add a tag specific for user metrics?

Reply:
Ok, added.